platform/search/src/services/GenericSearchProvider.js

```diff
@@ -171,17 +174,29 @@ define([
     GenericSearchProvider.prototype.index = function (id, model) {
         var provider = this;
 
-        this.worker.postMessage({
-            request: 'index',
-            model: model,
-            id: id
+        if (id !== 'ROOT') {
+            this.worker.postMessage({
+                request: 'index',
+                model: model,
+                id: id
+            });
+        }
+
+        var domainObject = objectUtils.toNewFormat(model, id);
+        var composition = _.find(this.openmct.composition.registry, function (p) {
+            return p.appliesTo(domainObject);
         });
 
-        if (Array.isArray(model.composition)) {
-            model.composition.forEach(function (idToIndex) {
-                provider.scheduleForIndexing(idToIndex);
-            });
+        if (!composition) {
+            return;
         }
+
+        composition.load(domainObject)
+            .then(function (children) {
+                children.forEach(function (child) {
+                    provider.scheduleForIndexing(objectUtils.makeKeyString(child));
```

Comment:
Echoing the `modelService` comment above, this seems like a missed opportunity to move away from legacy APIs. We have the child models that we want to index in this loop, but instead they get turned into key-strings and are later looked up from the `modelService`.

On the other hand, I'm also seeing how this would be a broader set of changes to GenericSearch, which has undergone a lot of performance-tuning and debugging. Maybe better to file a refactor issue for later to focus on those concerns more narrowly?